File: xdm-app/src/main/java/xdm/app/ui/components/DownloadSorter.java
package xdm.app.ui.components;

import xdm.app.AppContext;
import xdm.app.models.DownloadEntry;

import java.util.Comparator;

public class DownloadSorter implements Comparator<DownloadEntry> {

  @Override
  public int compare(DownloadEntry o1, DownloadEntry o2) {
    var sortKey = AppContext.INSTANCE.getConfigService().getSortKey();
    var ascending = AppContext.INSTANCE.getConfigService().isSortAscending();
    int res = 0;
    switch (sortKey) {
      case NAME: // sort by name
        res = o1.getFileName().compareTo(o2.getFileName());
        break;
      case SIZE: // sort by size
        res = Long.compare(o1.getSize(), o2.getSize());
        break;
      case DATE: // sort by date
        res = Long.compare(o1.getDateEpoch(), o2.getDateEpoch());
        break;
      case TYPE: // sort by type
        res = o1.getState().compareTo(o2.getState());
        break;
      default:
        break;
    }
    return ascending ? -res : res;
  }
}

File: xdm-app/src/main/java/xdm/app/ui/components/DownloadSorter.kt
package xdm.app.ui.components

import xdm.app.AppContext.config
import xdm.app.constants.SortKey
import xdm.app.data.DbRecord

class DownloadSorter : Comparator<DbRecord> {
    override fun compare(o1: DbRecord, o2: DbRecord): Int {
        val sortKey = config.sortKey
        val ascending = config.isSortAscending
        var res = 0
        when (sortKey) {
            SortKey.NAME -> res = o1.fileName.compareTo(o2.fileName)
            SortKey.SIZE -> res = o1.size.compareTo(o2.size)
            SortKey.DATE -> res = o1.date.compareTo(o2.date)
            SortKey.TYPE -> res = o1.status.compareTo(o2.status)
            else -> {
                // No ops
            }
        }
        return if (ascending) -res else res
    }
}
